community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/DocValuesCollector.java

```diff
@@ -0,0 +1,852 @@
+     * Two, lookup a value for the current doc in a sidecar {@code NumericDocValues} field.
+     * That is, this iterator has a main field, that drives the iteration and allow for lookups
+     * in other, secondary fields based on the current document of the main iteration.
+     *
+     * Lookups from this class are not thread-safe. Races can happen when the segment barrier
+     * is crossed; one thread might think it is reading from one segment while another thread has
+     * already advanced this Iterator to the next segment, having raced the first thread.
+     */
+    public class LongValuesIterator extends PrimitiveLongCollections.PrimitiveLongBaseIterator implements DocValuesAccess
+    {
+        private final Iterator<DocValuesCollector.MatchingDocs> matchingDocs;
+        private final String field;
+        private final int size;
+        private DocIdSetIterator currentDisi;
+        private NumericDocValues currentDocValues;
+        private DocValuesCollector.MatchingDocs currentDocs;
+        private final Map<String,NumericDocValues> docValuesCache;
+
+        private int index = 0;
+
+        /**
+         * @param allMatchingDocs all {@link DocValuesCollector.MatchingDocs} across all segments
+         * @param totalHits the total number of hits across all segments
+         * @param field the main field, whose values drive the iteration
+         */
+        public LongValuesIterator( Iterable<DocValuesCollector.MatchingDocs> allMatchingDocs, int totalHits, String field )
+        {
+            this.size = totalHits;
+            this.field = field;
+            matchingDocs = allMatchingDocs.iterator();
+            docValuesCache = new HashMap<>();
+        }
+
+        /**
+         * @return the number of docs left in this iterator.
+         */
+        public int remaining()
+        {
+            return size - index;
+        }
+
+        @Override
+        public long current()
+        {
+            return next;
+        }
+
+        @Override
+        public long getValue( String field )
+        {
+            if ( ensureValidDisi() )
+            {
+                if ( docValuesCache.containsKey( field ) )
+                {
+                    return docValuesCache.get( field ).get( currentDisi.docID() );
+                }
+
+                NumericDocValues docValues = currentDocs.readDocValues( field );
+                docValuesCache.put( field, docValues );
+
+                return docValues.get( currentDisi.docID() );
+            }
+            else
+            {
+                // same as DocValues.emptyNumeric()#get
+                // which means, getValue carries over the semantics of NDV
+                // -1 would also be a possibility here.
+                return 0;
+            }
+        }
+
+        @Override
+        protected boolean fetchNext()
+        {
+            try
+            {
+                if ( ensureValidDisi() )
```

Comment:
that is commonly used abbreviation for DocumentIdSetIterator, will rename it to avid confusion